Exercise-02/Exercise02/main.swift

```diff
@@ -52,18 +206,366 @@ func runApplication() {
+  var count = 0
+  
+  if let cityInfo = storedData[avgInfo.city] {
+    let daysBetween = Int(avgInfo.to.timeIntervalSince(avgInfo.from)) / (24 * 60 * 60) + 1
+    var daysToCheck: [DayInfo] = []
+    if daysBetween > cityInfo.dateToDayInfo.count {
+      // iterate over data and check if day is in interval
+      for (date, info) in cityInfo.dateToDayInfo {
+        if (date >= avgInfo.from && date <= avgInfo.to) {
+          daysToCheck.append(info)
+        }
+      }
+    } else {
+      // iterate over interval and check if day is in data
+      var currentDate = avgInfo.from
+      while currentDate <= avgInfo.to {
+        guard let nextDate = getCalendar().date(byAdding: .day, value: 1, to: currentDate) else {
+          throw TemperatureError.validationFailed("invalid date")
+        }
+        
+        if let dayInfo = cityInfo.dateToDayInfo[currentDate] {
+          daysToCheck.append(dayInfo)
+        }
+        currentDate = nextDate
+      }
     }
+    
+    for dayInfo in daysToCheck {
+      sum += dayInfo.sum()
+      count += dayInfo.count()
+    }
+  }
+  
+  guard count > 0 else {
+    throw TemperatureError.noData("no data for \(avgInfo.city) in given interval")
+  }
+  
+  let avg = sum / Double(count)
+  let fromStr = getDateStr(avgInfo.from)
+  let toStr = getDateStr(avgInfo.to)
+  print("\nAverage temperature for \(avgInfo.city) between \(fromStr) and \(toStr) is \(formatTemperature(celsiusTemperature: avg, unit: avgInfo.unit))")
+}
+
+func getUnitOfMeasure(commandSplit: [String.SubSequence], unitOfMeasureIndex: Int) throws -> UnitOfMeasure {
+  guard let unitOfMeasure = UnitOfMeasure(rawValue: unitOfMeasureIndex < commandSplit.endIndex ? String(commandSplit[unitOfMeasureIndex]) : UnitOfMeasure.celsius.rawValue) else {
+    throw TemperatureError.invalidInput("invalid temperature unit")
+  }
+  
+  return unitOfMeasure
+}
+
+func getDateAndTime(_ dateAndTimeStr: String) throws -> Date {
+  let dateAndTimeFormat = "yyyy-MM-dd'T'HH:mm:ss"
+  let dateAndTimeFormatter = DateFormatter()
+  dateAndTimeFormatter.locale = Locale(identifier: "en_US_POSIX")
+  dateAndTimeFormatter.dateFormat = dateAndTimeFormat
+  dateAndTimeFormatter.timeZone = TimeZone(secondsFromGMT: 0)
+  
+  guard let dateAndTime = dateAndTimeFormatter.date(from: dateAndTimeStr) else {
+    throw TemperatureError.invalidInput("invalid date and time, format should be \(dateAndTimeFormat)")
+  }
+  
+  return dateAndTime
+}
+
+func getDate(_ dateStr: String) throws -> Date {
+  let dateFormat = "yyyy-MM-dd"
+  let dateFormatter = DateFormatter()
+  dateFormatter.locale = Locale(identifier: "en_US_POSIX")
+  dateFormatter.dateFormat = dateFormat
+  dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
+  
+  
+  guard let date = dateFormatter.date(from: dateStr) else {
+    throw TemperatureError.invalidInput("invalid date, format should \(dateFormat)")
   }
+  
+  return date
 }
 
-// TODO: Add your functions
+func getDateStr(_ date: Date) -> String {
```

Comment:
```suggestion
func getDateString(_ date: Date) -> String {
```
Don't shorten names.